### PR TITLE
refactor(yaml): replace `Type` class with interface

### PR DESCRIPTION
**Pull request by timreichen**
- removes `Type` class and replaced with a `Type` interface.

### CODE REVIEW

**Review by iuioiua — APPROVED** [review]
I like this and was going to do this myself at some point. LGTM.

File: yaml/_type.ts [hunk 1]
```diff
@@ -10,44 +10,15 @@ export type KindType = "sequence" | "scalar" | "mapping";
...
-      this.instanceOf = options.instanceOf;
-      this.predicate = options.predicate;
-      this.represent = options.represent;
-      this.defaultStyle = options.defaultStyle;
-      this.styleAliases = options.styleAliases;
-    }
-  }
-  resolve: (data?: Any) => boolean = (): boolean => true;
-  construct: (data?: Any) => Any = (data): Any => data;
+  resolve?: (data?: Any) => boolean;
```
Comment (kt3k) [inline yaml/_type.ts:22]: Making resolve optional feels strange to me

File: yaml/_type.ts [hunk 2]
```diff
@@ -10,44 +10,15 @@ export type KindType = "sequence" | "scalar" | "mapping";
...
+  construct?: (data?: Any) => Any;
```
Comment (kt3k) [inline yaml/_type.ts:23]: I don't think this should be optional

**Comment by kt3k** [comment]
I'm not in favor of this change. There are many changes in optionality of fields. I'm very unsure this doesn't introduce any bugs.

**Review by iuioiua — COMMENTED** [review]
I think this change is an improvement in design. It also seems safe enough to do.

However, I understand Yoshiya's sentiment. In #5239, I made changes which I thought were harmless, as tests passed. But rather testing is not yet at a level for this package to instil such confidence.

@kt3k, if you feel strongly about this, let's close this for now and revisit once testing is at an adequate level.

File: yaml/_loader/loader.ts [hunk 3]
```diff
@@ -1545,14 +1547,16 @@ function composeNode(
         );
       }
 
-      if (!type.resolve(state.result)) {
+      if (!type.resolve?.(state.result) ?? true) {
```
Comment (kt3k) [inline yaml/_loader/loader.ts:1550]: This condition is reversed when `resolve` is a default implementation

File: yaml/_type.ts (hunk 1 above)
Comment (kt3k) [reply yaml/_type.ts:22]: > An alternative would be to require resolve and construct methods in Type and add them to types that are missing them.

I think that is more correct way to perform this transition.

Comment (kt3k) [reply yaml/_type.ts:22]: this line https://github.com/denoland/deno_std/pull/5262/files#r1663457908 seems a bug caused by the optionality of the field

File: yaml/_loader/loader.ts (hunk 3 above)
Comment (iuioiua) [reply yaml/_loader/loader.ts:1550]: Nice catch! Yeah, if we're to go through with this PR, let's make `Type.resolve()` and `Type.construct()` required.

File: yaml/_type.ts (hunk 2 above)
Comment (timreichen) [reply yaml/_type.ts]: Reverted.

File: yaml/_loader/loader.ts (hunk 3 above)
Comment (timreichen) [reply yaml/_loader/loader.ts]: Done.

**Review by kt3k — APPROVED** [review]
LGTM

**Pull request merged**